lib/sycamore/sycamore/connectors/weaviate/weaviate_reader.py

```diff
@@ -0,0 +1,93 @@
+from sycamore.data import Document
+from sycamore.connectors.common import unflatten_data
+from sycamore.connectors.base_reader import BaseDBReader
+from dataclasses import dataclass
+from typing import Optional, Dict, Any
+from dataclasses import asdict
+from weaviate.client import (
+    AdditionalConfig,
+    AuthCredentials,
+    ConnectionParams,
+    EmbeddedOptions,
+)
+from weaviate import WeaviateClient
+
+
+@dataclass
+class WeaviateReaderClientParams(BaseDBReader.ClientParams):
+    connection_params: Optional[ConnectionParams] = None
+    embedded_options: Optional[EmbeddedOptions] = None
+    auth_client_secret: Optional[AuthCredentials] = None
+    additional_headers: Optional[dict] = None
+    additional_config: Optional[AdditionalConfig] = None
+    skip_init_checks: bool = False
+
+
+@dataclass
+class WeaviateReaderQueryParams(BaseDBReader.QueryParams):
+    collection_name: str
+    query_kwargs: Optional[Dict] = None
+
+
+class WeaviateReaderClient(BaseDBReader.Client):
+    def __init__(self, client: WeaviateClient):
+        self._client = client
+
+    @classmethod
+    def from_client_params(cls, params: BaseDBReader.ClientParams) -> "WeaviateReaderClient":
+        assert isinstance(params, WeaviateReaderClientParams)
+        client = WeaviateClient(**asdict(params))
+        return WeaviateReaderClient(client)
+
+    def read_records(self, query_params: BaseDBReader.QueryParams) -> "WeaviateReaderQueryResponse":
+        assert isinstance(
+            query_params, WeaviateReaderQueryParams
+        ), f"Wrong kind of query parameters found: {query_params}"
+        with self._client:
+            query = None
+            collection = []
+            if query_params.query_kwargs:
+                query = self._client.collections.get(query_params.collection_name).query
+                for method, value in query_params.query_kwargs.items():
+                    if value is not None:
+                        method_name = f"{method}"
+                        if hasattr(query, method_name):
+                            query = getattr(query, method_name)(**value)
+                        else:
+                            raise ValueError(f"Error: Method '{method_name}' not found in query object.")
+                query = query.objects  # type: ignore
+            else:
+                collection = list(
+                    self._client.collections.get(query_params.collection_name).iterator(include_vector=True)
+                )
+            results = WeaviateReaderQueryResponse(collection=collection, query_output=query)
```

Comment:
I think this is genius, which makes me want to see a bunch of unit tests showing various cases

Reply:
Just added